src/lib/components/Table/Table.js

```diff
@@ -0,0 +1,199 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import naturalCompare from 'natural-compare-lite';
+import './Table.scss';
+
+import TableCell from './TableCell';
+import TableRow from './TableRow';
+
+class Table extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.generateData = this.generateData.bind(this);
+    this.generateHeader = this.generateHeader.bind(this);
+    this.generateRows = this.generateRows.bind(this);
+    this.getNewSortCondition = this.getNewSortCondition.bind(this);
+    this.sortData = this.sortData.bind(this);
+
+    this.state = {
+      columns: props.columns ? [...props.columns] : null,
+      data: props.data ? [...props.data] : null,
+      originalData: props.data ? [...props.data] : null,
+      sortCondition: props.sortCondition,
+    };
+  }
+
+  componentWillMount() {
+    const { columns, data, sortCondition } = this.state;
+
+    // If Table is static, convert data to object
+    if (!data || !columns) {
+      this.generateData();
+    }
+
+    // Pre-sort Table if sortCondition prop is non-zero
+    if (Object.keys(sortCondition).length === 1) {
+      const columnKey = Object.keys(sortCondition)[0];
+      this.sortData(columnKey);
+    }
+  }
+
+  getNewSortCondition(columnKey) {
+    const { sortCondition } = this.state;
+    const newSortCondition = {};
+    const options = ['ascending', 'descending', 'none'];
+
+    // Move to next option in options array
+    newSortCondition[columnKey] = (
+      options[(options.indexOf(sortCondition[columnKey]) + 1) % options.length]
+    );
+
+    this.setState({ sortCondition: newSortCondition }, () => {
+      this.sortData(columnKey);
+    });
+  }
+
+  sortData(columnKey) {
+    const { data, sortCondition } = this.state;
+
+    if (sortCondition[columnKey] === 'none') {
+      const unsortedData = this.state.originalData;
+      this.setState({ data: [...unsortedData] });
+    } else {
+      const sortedData = data.sort((a, b) => {
+        const x = a[columnKey];
+        const y = b[columnKey];
+        return (
+          sortCondition[columnKey] === 'ascending' ? naturalCompare(x, y) : naturalCompare(y, x)
+        );
+      });
+
+      this.setState({ data: sortedData });
+    }
+  }
+
+  generateData() {
+    const { children } = this.props;
+    const columns = [];
+    const data = [];
+
+    children.forEach((row, rowIndex) => {
+      if (row.props.tableHeader) {
+        row.props.children.forEach((cell, cellIndex) => {
+          let value = '';
+          if (typeof cell.props.children === 'string') value = cell.props.children;
```

Comment:
@Lukewh typescript ....

Reply:
@tbille yeah I'll learn it eventually...